### ISSUE TITLE
Filesystem based routing?

### ISSUE THREAD

**Issue by m1212e** [elysiajs/elysia#65]
Hi,

I recently got into bun and was looking for a nice webframework which supports filesystem based routing. I like it because it - by design - enforces a clear structure for the project and you know where to find a certain route. Elysia does a few other things really really well in my opinion so I thought about using it but as far as I see it does not support fs based routes. Is this a thing you considered? Maybe also in a hybrid approach like I did here: https://github.com/m1212e/bun-fs-router-plugin

**Comment by itsyoboieltr** [comment]
+1, I would be really interested in this, I was just about to raise a proposal for this exactly.

**Comment by Fecony** [comment]
But Elysia doesn't really have frontend part so file based routing for api doesn't make sense..?

**Comment by itsyoboieltr** [comment]
> But Elysia doesn't really have frontend part so file based routing for api doesn't make sense..?

This has nothing to do with frontend or not.

For reference, check [Next.js route handler](https://nextjs.org/docs/app/building-your-application/routing/router-handlers) implementation for a file-system based api.

**Comment by Fecony** [comment]
> > But Elysia doesn't really have frontend part so file based routing for api doesn't make sense..?
> 
> This has nothing to do with frontend or not.
> 
> For reference, check [Next.js route handler](https://nextjs.org/docs/app/building-your-application/routing/router-handlers) implementation for a file-system based api.

I didn't play with file based api, from one side looks like overhead of splitting everything
on the other looks like a good solution to extract those and make smaller

**Comment by ethanniser** [comment]
I don't think it would be too hard to make your own simple implementation
See the "*However, it isn't too difficult to make a file-based router. Here's a file-based example code with builder:*" part of the [Waku docs](https://waku.gg/docs/practices/router)

**Comment by wobsoriano** [comment]
Here's one - https://github.com/wobsoriano/elysia-autoroutes

**Comment by itsyoboieltr** [comment]
I also made an implementation of a file system routing plugin, based on groups, called the [group router](https://github.com/itsyoboieltr/elysia-group-router).

**Comment by m1212e** [comment]
There are a few options, you could calculate the routes based on the current file path and just pass a helper function to spit out the correct relative path based on the file location, you could also utilize the bun.js macros/build plugins to set the correct path at build time. But what I could not get to work properly was the eden type safety. Since it work without code generation and TS does not support build time macros, I don't really seee a way to pull this off. Maybe one of the elysia devs knows how to implement something like this.

**Comment by ethanniser** [comment]
was just randomly reading the bun docs today and saw it has a native `FileSystemRouter` API that may be of use here

https://bun.sh/docs/api/file-system-router

**Comment by wobsoriano** [comment]
@ethanniser https://github.com/wobsoriano/elysia-autoroutes uses it!

**Comment by wobsoriano** [comment]
> was just randomly reading the bun docs today and saw it has a native `FileSystemRouter` API that may be of use here
> 
> https://bun.sh/docs/api/file-system-router

[elysia-autoroutes](https://github.com/wobsoriano/elysia-autoroutes/blob/09f6fd73d024380e035c48c086100f33d14d47b6/src/utils/autoload.ts#L19) uses it!

**Comment by m1212e** [comment]
@wobsoriano Did you manage to get eden types to work with it?

**Comment by wobsoriano** [comment]
No. That's the biggest challenge @m1212e - how do we combine types from the main ElysiaApp instance with the route files?

**Comment by m1212e** [comment]
I don't see a way to do this without running some kind of process at development time.

**Comment by wobsoriano** [comment]
@m1212e what other meta framework do, like Nuxt, is to generate types for each route at dev time

**Comment by itsyoboieltr** [comment]
> @m1212e what other meta framework do, like Nuxt, is to generate types for each route at dev time

@wobsoriano Actually, I tried to do something similar here (with not much success): https://github.com/elysiajs/elysia/issues/79
